[docs] document new observe features (#45250)
expo/expo: docs/pages/eas/observe/configuration.mdx

To flush events manually (for example, during testing or to ensure events are sent before a specific point), call `dispatchEvents()`:

```tsx
import ExpoObserve from 'expo-observe';

await ExpoObserve.dispatchEvents();
```

## Sampling

By default, every installation dispatches its metrics. For high-volume apps, you can sample a fraction of installations by setting `sampleRate` to a value between `0` and `1`:

```tsx
import ExpoObserve from 'expo-observe';

// Dispatch metrics from ~25% of installations.
ExpoObserve.configure({
  sampleRate: 0.25,
});
```

The sampling decision is **deterministic per installation**. Each installation is either permanently in-sample or out-of-sample for a given rate, so the choice is stable across app launches and you get a consistent slice of installations rather than a random subset of sessions.

A few details worth knowing:

- Values outside the `[0, 1]` range are clamped to the nearest edge. `0` always drops; `1` always dispatches.
- Out-of-sample devices drop pending metrics rather than accumulating them. Lowering the rate later does not retroactively send earlier sessions.

> - Sampling is independent of [`dispatchingEnabled`](#configure). If `dispatchingEnabled` is `false`, nothing is dispatched regardless of `sampleRate`.

Reviewer: ```suggestion
- Sampling depends on [`dispatchingEnabled`](#configure). If `dispatchingEnabled` is `false`, nothing is dispatched regardless of `sampleRate`.
```

await ExpoObserve.dispatchEvents();
```

## Sampling

By default, every installation dispatches its metrics. For high-volume apps, you can sample a fraction of installations by setting `sampleRate` to a value between `0` and `1`:

> ```js

Reviewer 2: ````suggestion
```ts
````

import ExpoObserve from 'expo-observe';

// Dispatch metrics from ~25% of installations.
ExpoObserve.configure({
  sampleRate: 0.25,
});
```

The sampling decision is **deterministic per installation**. Each installation is either permanently in-sample or out-of-sample for a given rate, so the choice is stable across app launches and you get a consistent slice of installations rather than a random subset of sessions.

A few details worth knowing:

> - Values outside the `[0, 1]` range are clamped. `0` always drops; `1` always dispatches.

Reviewer 3: [🐧 nit] Small precision suggestion: "Values outside the `[0, 1]` range are clamped." doesn't say *to what*. "Values outside `[0, 1]` are clamped to the nearest edge" (or to `0`/`1` respectively) makes it unambiguous, especially since the next half of the sentence then talks about the `0` and `1` boundary cases.

> - Values outside the `[0, 1]` range are clamped. `0` always drops; `1` always dispatches.

Author: ```suggestion
- Values outside the `[0, 1]` range are clamped to the nearest edge. `0` always drops; `1` always dispatches.
```

- Out-of-sample devices drop pending metrics rather than accumulating them. Lowering the rate later does not retroactively send earlier sessions.
- Sampling is independent of [`dispatchingEnabled`](#configure). If `dispatchingEnabled` is `false`, nothing is dispatched regardless of `sampleRate`.

## Enable metrics in development

By default, metrics collected from debug builds are not dispatched. To dispatch them anyway (for example, while testing your Expo Observe integration), set `dispatchInDebug` to `true` when calling `configure()`:

> ```js

Reviewer 2: ```suggestion
```ts
```

import ExpoObserve from 'expo-observe';

ExpoObserve.configure({
  dispatchInDebug: true,
});
```

`dispatchInDebug` has no effect on release builds, which always dispatch (subject to [`dispatchingEnabled`](#configure) and [`sampleRate`](#sampling)). If `dispatchingEnabled` is `false` or this installation is out-of-sample, nothing dispatches regardless of `dispatchInDebug`.

> > **warning** Enable this only when testing your Expo Observe integration. Development performance differs significantly from production, so collecting development metrics may distort the results shown in your dashboard.

Reviewer 3: [🐧 nit] Minor wording carryover: "Development performance differs significantly from production, so collecting development metrics may distort the results..." — the rest of this section now talks about *debug builds*, but this callout still talks about "development" performance and "development metrics". For consistency, consider "Debug-build performance differs significantly from production, so collecting debug-build metrics may distort the results shown in your dashboard."

Reviewer 2: Or use both (development/debug) for clarity. (just a suggestion).

What do you use for reviews like this, @tsapeta, if you don't mind me asking?

Reviewer 3: @amandeepmittal @kadikraman It's Kudo's skill, available here: https://gist.github.com/Kudo/9b531c5bbd573e2d2c2434cd7a9b8ae3

Reviewer 2: Amazing! Thanks @tsapeta!

> > **warning** Enable this only when testing your Expo Observe integration. Development performance differs significantly from production, so collecting development metrics may distort the results shown in your dashboard.

Author: ```suggestion
> **warning** Enable this only when testing your Expo Observe integration. Development/debug performance differs significantly from production, so collecting development/debug metrics may distort the results shown in your dashboard.
```
